---
title: "Analysis of Single-Cell RNA-seq - Session 4"
author: "Emily V Chambers"
date: '`r format(Sys.time(), "Last modified: %d %b %Y")`'
output: 
  html_notebook: 
    toc: yes
    toc_float: yes
    css: stylesheets/styles.css
---
<img src="images/logo-sm.png" style="position:absolute;top:40px;right:10px;" width="200" />

```{r setup, include=FALSE}
knitr::opts_chunk$set(
  tidy = TRUE,
  tidy.opts = list(width.cutoff = 95),
  message = FALSE,
  warning = FALSE,
  fig.width = 10,
  time_it = TRUE
)

```

# Introduction

To re-cap -  we have obtained our raw counts matrices from GEO (or cellranger). We have performed QC and subset the data according to the QC plots we generated. We have normalised the data, selected the most variable features and scaled - regressing out unwanted sources of variation. 

In the next part of the course we are going to explore differentially expressed genes between clusters and assigning cluster identity. 

## Learning Objectives
* Find differentially expressed genes using `FindMarkers()` and `FindAllMarkers()`.
* Use the biocondictor human annotation package `org.Hs.eg.db` to add some more further annotation to our marker genes.
* Assign cluster identity using canonical marker genes
* Carry out a simple pathway analysis

![](images/flowchart2.png)

Firstly, we will load up the packages we will needfor this part of the course.

```{r}
library(Seurat, quietly = T)
library(tidyverse, quietly = T)
library(org.Hs.eg.db, quietly = T)
```
And then we can load up the R objects we have used in the previous sessions.
```{r}
seurat <- readRDS("Robjects/seurat_integrated.RDS")
```

We can simple run the name of the Seurat object (in our case 'seurat') to give us a summary of what the object contains and what calculations have been performed on it.
```{r, eval=FALSE, echo=F}
seurat
```

We can recap our clustering by taking a look at the `DimPLot`
```{r}
DimPlot(seurat, reduction = "umap")
```


# Find differentially expressed genes
Our next objective is to start assigning some **biological meaning** to our data. We have some nice clusters but no clue yet what cell groups out clusters contain.

In Seurat we can find the top differentially expressed genes, or markers, within each cluster. The function `FindMarkers()` handles most of the differential expression testing within the Seurat package and finds marker genes within a particular cluster. As a default, Seurat performs differential expression based on the non-parameteric Wilcoxon rank sum test but other methods are available within the function. 

To start with we can find all the differentially expressed genes in cluster 8 compared to every other cluster. The min.pct argument requires a gene to be detected at a minimum percentage of cells.

```{r}
# We make sure the default assay is set back to "RNA" to use the original counts for DE analysis.
DefaultAssay(seurat) <- "RNA"
```


```{r}
cluster3.markers <- FindMarkers(seurat, ident.1 = 3, min.pct  = 0.25, verbose = F)
cluster3.markers %>% head
```
A quick look at the top 6 lines in our differentially expressed marker gene table shows us the format of the data. We have the results of the Wilcoxon rank sum test as a p value and adjusted p value for multiple testing. We also have our average log2 fold change in expression between our clusters. The pct.1 and pct.2 columns show us the percentage of cells within the cluster the gene is expressed in (we set a minimum threshold of 20% in the function). As default, the output table is sorted by adjusted p value.

A little  bit of table manipulation in `dplyr` makes the markers easier to view...
```{r}
cluster3.markers  %>% dplyr::select(avg_log2FC, p_val, p_val_adj) %>%slice_max(n = 5, order_by = avg_log2FC)
```
To check our DE findings, we can visualize expression of a particular gene of interest across all clusters using the `VlnPlot()` function. This shows expression of a gene (or genes) of interest across all clusters.

```{r}
VlnPlot(seurat, features = "LYZ")
```
We can also check which cells express out gene of interest on the UMAP image using `FeaturePlot`

```{r}
FeaturePlot(seurat, features = "LYZ")

```
Perhaps we are most interested in the differentially expressed genes between clusters 1 and 3

```{r}
cluster1.markers <- FindMarkers(seurat, ident.1 = 1, ident.2 = 3, min.pct = 0.25, verbose = F)
cluster1.markers %>% head()
```
... or the DE genes between clusters 1 in comparison to 3 and 4?
```{r}
cluster1.markers <- FindMarkers(seurat, ident.1 = 1, ident.2 = c(3,4), min.pct = 0.25, verbose = F)
cluster1.markers %>% head()
```
Usually, we are wanting to find the most differentially expressed genes in each cluster across the whole dataset. For this we use `FindAllMarkers`. This time we will only report the positively expressed markers with a minimum log fold change of 0.25.

```{r}
markers <- FindAllMarkers(seurat, only.pos = TRUE, min.pct = 0.25, logfc.threshold = 0.25, verbose = F)

markers %>%
  dplyr::select(gene, cluster, avg_log2FC, p_val, p_val_adj) %>%
    group_by(cluster) %>%
    slice_max(n = 5, order_by = avg_log2FC)
```


## Exercise

<div class="exercise">
- Use `FindMarkers` to find the differentially expressed genes in clusters 1 & 2 vs 8
- Can you use `dplr` to select columns of interest and order them by log fold change?
- Use one of the visualisation tools to look at the expression of one of the top marker genes
</div>

```{r}
#Put exercise work in here :-)


```


# Annotating marker genes

We can use gene databases within R to further annotate our marker genes of interest. The `org.Hs.eg.db` database is a genome wide annotationpackage for human that has many different gene mappings based on Entrez Gene identifiers.

In our case we are going to use the gene symbols (rownames in our Seurat object) to extract the gene description and ensembl ID from the database.

```{r}
columns(org.Hs.eg.db) # We can use this command to see which data can be pulled from the database

anno <- AnnotationDbi::select(org.Hs.eg.db,keys=rownames(seurat),columns=c("SYMBOL","GENENAME", "ENTREZID"), keytype = "SYMBOL")

# Check it worked!
head(anno)  
```
Now we can combine our annotation table to our marker genes using `dplyr` and `left_join`
```{r}
anno_markers <- markers %>% 
  dplyr::select(gene, cluster, avg_log2FC, p_val, p_val_adj) %>%
  left_join(anno, by = c("gene" = "SYMBOL"))

anno_markers %>% head()

```
# Assigning cluster identity

To assign biological meaning to our clusters, we are going to use canonical cell type markers from the paper. These genes are known to be expressed in a particular cell type.

Gene    Cell type
----    ----
CD34    CD79A B cells, 
MS4A1   CD20+ B cells, 
CST3    Monocytes, 
SPN   Immature hematopoietic, 
HBA1    erythrocytes, 
CD3D    T cells, 
NKG7    NK cells, 
MZB1    BCMA

We can assign these marker genes as a variable
```{r}
cell_markers <- c("CD34", "MS4A1", "CST3" ,"SPN", "HBA1", "CD3D", "NKG7",  "MZB1")
```

Dot plots allow us to examine particular features or genes of interest across all clusters. Creating a dotplot of the cell type markers lets us assign identity based on their expression.
```{r}
DotPlot(seurat, features = cell_markers) + theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))
```
We can then manually create new identities for each cluster as follows.

```{r}
new.cluster.ids <- c("T Cells", "T Cells", "Erythrocytes",  "Monocytes", "x", "CD79A B cells", "BCMA1", "NK Cells", "x2", "BCMA", "Monocytes", "CD20+ B cells", "BCMA", "Erythrocytes", "BCMA")
names(new.cluster.ids) <- levels(seurat)
new.cluster.ids
seurat_named <- RenameIdents(seurat, new.cluster.ids)
```

```{r}
DimPlot(seurat_named, reduction = "umap", label = TRUE, pt.size = 0.5) + NoLegend()
```

# GO term and pathway enrichment

Explanatory text

```{r}
library(clusterProfiler)
background <- anno %>% pull("ENTREZID")

go_markers <- anno_markers %>% filter(cluster ==1) %>% pull(ENTREZID)

  eGO <- enrichGO(
    gene          = go_markers,
    OrgDb         = org.Hs.eg.db,
    ont           = "BP",
    pAdjustMethod = "BH",
    pvalueCutoff  = 0.05,
    readable      = TRUE
  )
  
  eGO[]

eGO[] %>%
  slice_max(n = 10, order_by = -log10(p.adjust)) %>% 
  ggplot(aes(x = -log10(p.adjust), y = Description)) +
  geom_col(fill="blue") + 
  xlab('-log10 FDR') + 
  ylab(NULL)



```



