---
title: "R Notebook"
output: html_notebook
---

# Introduction

In this workshop we will be using the dataset:-

> Caron M, St-Onge P, Sontag T, Wang YC et al. Single-cell analysis of childhood leukemia reveals a link between developmental states and ribosomal protein expression as a source of intra-individual heterogeneity. Sci Rep 2020 May 15;10(1):8079

The *processed* single-cell data have been made available on Gene Expression Omnibus (GEO) and we have downloaded them for you. For more information, the GEO entry is [here](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE132509) and the code to download is given in the R script [download_GSE132509.R](download_GSE132509.R).

*We will only analyse a subset of the dataset during the practical*.

```{r}
library(Seurat)
library(tidyverse)
```

# Data Import

We start by importing the sample information for the dataset. If you are interested, the creation of this file is shown in [download_GSE132509.R](download_GSE132509.R). The sampleSheet data frame contains information about each of the samples in the dataset; which GEO accession they correspond to and the disease status of the sample.

```{r}
sampleSheet <- read.delim("GSE132509/meta_data.tsv")
sampleSheet
```

## Code to import a single biological sample

The code to read a single sample is given below. This assumes that the output from the *cellranger* tool can be found in a specified directory. As part of the course setup ([`download_GSE132509.R`]("download_GSE132509.R")) we have already renamed the outputs to the convention that `Seurat` is expecting. 

```{r}
list.files("GSE132509/GSM3872434/")
```

It is then a two-step process to create a object that `Seurat` can use for analysis.

```{r}
seurat_data <- Read10X(data.dir = "GSE132509/GSM3872434")
seurat_obj <- CreateSeuratObject(counts = seurat_data, 
                                 min.features = 100, 
                                 project = "ETV6-RUNX1_1")
seurat_obj
```

Note that the "samples" referred to in the output are individual cells that have been sequenced. We can inspect the data that we have just imported. The counts can be displayed, although most of them will be empty due to the *sparse* nature of single-cell data.

```{r}
GetAssayData(seurat_obj, "counts")[1:10,1:2]
```

The "meta-data" in the object is used to record any QA data that we generated; used to filter any poor-quality cells from the dataset. We will see how to add to this data frame later.

```{r}
seurat_obj@meta.data %>% head


```

## 

```{r}
seurat_data <- Read10X(data.dir = "GSE132509/GSM3872442")
seurat_obj2 <- CreateSeuratObject(counts = seurat_data, 
                                 min.features = 100, 
                                 project = "PBMMC_1")
seurat_obj2
merged_data <- merge(x = seurat_obj,y=seurat_obj2)
```



## Importing many samples

We will now employ a "for loop" to process the remainder of the samples. In programming, looping is a strategy for repeating same lines of code but for a set of different inputs. To see why this is beneficial, lets look at the code for importing the second and third samples in our dataset:-

```{r eval=FALSE}
## No need to run this code!
## It is just for an example

seurat_data <- Read10X(data.dir = "GSE132509/GSM3872435")
seurat_obj <- CreateSeuratObject(counts = seurat_data, 
                                 min.features = 100, 
                                 project = "ETV6-RUNX1_2")
seurat_obj

seurat_data <- Read10X(data.dir = "GSE132509/GSM3872436")
seurat_obj <- CreateSeuratObject(counts = seurat_data, 
                                 min.features = 100, 
                                 project = "ETV6-RUNX1_3")
seurat_obj
```

The only 


```{r eval=FALSE}
dirs <- slice(sampleSheet, c(2:4,9:11)) %>% 
        pull(GSE)

samples <- slice(sampleSheet, c(2:4,9:11)) %>% 
        pull(SampleName)
seurat_list <- NULL

# Create a Seurat object for each sample
for (i in 1:length(dirs)){
        message(paste("Reading data from ", dirs[i]))
        seurat_data <- Read10X(data.dir = paste0("GSE132509/",dirs[i]))
        seurat_list[[i]] <- CreateSeuratObject(counts = seurat_data, 
                                         min.features = 100, 
                                         project = samples[i])
        
}

merged_data <- merge(x = seurat_obj,y=seurat_list)

```

```{r}
rm(seurat_list)
rm(seurat_data)
rm(seurat_obj)
gc()
```


```{r}
merged_data[["mito.ratio"]] <- PercentageFeatureSet(merged_data, pattern="^MT")/100
merged_data@meta.data
```


```{r}
qc_data <- merged_data@meta.data %>% 
  data.frame %>% 
  left_join(sampleSheet, by = c("orig.ident"="SampleName")) %>% 
  rename(SampleName = orig.ident) %>% 
  rename(nUMI = nCount_RNA, nGene=nFeature_RNA)

View(qc_data)
```


## Number of cells per-sample

```{r}
qc_data %>% 
  ggplot(aes(x = SampleName, fill=SampleGroup)) + geom_bar() + theme(axis.text.x = element_text(angle=45,vjust=1,hjust=1))
```
## UMI counts per cell

```{r}
qc_data %>% 
  ggplot(aes( x= nUMI,fill=SampleName)) + geom_density(alpha=0.2) + scale_x_log10()
```

## Genes detected per cell

```{r}
qc_data %>% 
  ggplot(aes( x= nGene,fill=SampleName)) + geom_density(alpha=0.2)
```


## Complexity

Look at the relationship between the number of UMI and number of genes detected. Should be somewhat correlated

```{r}
ggplot(qc_data, aes(x = log10(nUMI), y= log10(nGene))) + geom_point() + geom_smooth(method="lm") + facet_wrap(~SampleName)
```

Define novelty score and add to the qc data

```{r}
qc_data <- mutate(qc_data, Novelty = log10(nGene) / log10(nUMI))

ggplot(qc_data, aes(x = log10(nUMI), y= log10(nGene),col=Novelty)) + geom_point() + facet_wrap(~SampleName)

```


## Mitochondrial counts

```{r}
qc_data %>% 
  ggplot(aes(x = mito.ratio, fill=SampleName)) + geom_density(alpha=0.2) + facet_wrap(~SampleGroup) + geom_vline(xintercept = 0.2)
```

## Considering multiple QC metrics

```{r}
qc_data %>% 
  ggplot(aes(x = nUMI, y = nGene, col=mito.ratio)) + geom_point() + facet_wrap(~SampleName) + geom_vline(xintercept = 500) + geom_hline(yintercept = 250)
```

```{r}
qc_data %>% 
  ggplot(aes(x = nUMI, y = nGene, col=Novelty)) + geom_point() + facet_wrap(~SampleName) + geom_vline(xintercept = 500) + geom_hline(yintercept = 250)
```


```{r}
qc_data %>% 
  mutate(kept_in_analysis = nUMI >= 500 & nGene >=250 & Novelty > 0.8 & mito.ratio < 0.2) %>%
  count(kept_in_analysis)
```

```{r}
qc_data %>% 
  mutate(kept_in_analysis = nUMI >= 500 & nGene >=250 & Novelty > 0.8 & mito.ratio < 0.2) %>%
  ggplot(aes(x = nUMI, y = nGene, col=kept_in_analysis)) + geom_point() + facet_wrap(~SampleName) + geom_vline(xintercept = 500) + geom_hline(yintercept = 250)
```

```{r}
### Put the new meta data we've created into the Seurat object

qc_data <- data.frame(qc_data)
rownames(qc_data) <- rownames(merged_data@meta.data)

merged_data <- AddMetaData(merged_data, qc_data)

filtered_data <- subset(x = merged_data, 
                         subset= (nUMI >= 500) & 
                           (nGene >= 250) & 
                           (Novelty > 0.80) & 
                           (mito.ratio < 0.20))
dim(filtered_data)
rm(merged_data)
```


## Count filtering

```{r}
# Extract counts
counts <- GetAssayData(object = filtered_data, slot = "counts")

# Output a logical matrix specifying for each gene on whether or not there are more than zero counts per cell
nonzero <- counts > 0

```

```{r}
# Sums all TRUE values and returns TRUE if more than 10 TRUE values per gene
keep_genes <- Matrix::rowSums(nonzero) >= 10

# Only keeping those genes expressed in more than 10 cells
filtered_counts <- counts[keep_genes, ]

filtered_data <- CreateSeuratObject(filtered_counts, meta.data = filtered_data@meta.data)
```

# Sources of Variation

```{r}
cc_genes <- read.delim("regev_lab_cell_cycle_genes.txt")[,1]
s_genes <- cc_genes[1:43]
g2m_genes <- cc_genes[44:97]
```

```{r}
filtered_data <- NormalizeData(filtered_data)
seurat_phase <- CellCycleScoring(filtered_data, 
                                 g2m.features = g2m_genes, 
                                 s.features = s_genes)

# View cell cycle scores and phases assigned to cells                                 
View(seurat_phase@meta.data)  
```




```{r}
# Identify the most variable genes
seurat_phase <- FindVariableFeatures(seurat_phase, 
                     selection.method = "vst",
                     nfeatures = 2000, 
                     verbose = FALSE)
		     
# Scale the counts
seurat_phase <- ScaleData(seurat_phase)
```

```{r}
# Perform PCA
seurat_phase <- RunPCA(seurat_phase)

# Plot the PCA colored by cell cycle phase
DimPlot(seurat_phase,
        reduction = "pca",group.by = "Phase")
```

```{r}
summary(seurat_phase$mito.ratio)
seurat_phase$mito.cat <- cut(seurat_phase$mito.ratio, breaks=c(-Inf,0.02971, 0.03951,0.05186,Inf),labels=c("Low","Medium","Medium High","High"))

DimPlot(seurat_phase,
        reduction = "pca",
        group.by= "SampleGroup")

```



```{r message=FALSE}
seurat_split <- SplitObject(seurat_phase, split.by = "SampleGroup")

for(i in 1:length(seurat_split)){
  
  seurat_split[[i]] <- SCTransform(seurat_split[[i]],vars.to.regress = c("mito.ratio","nUMI","S.Score","G2M.Score") )
}

```

```{r}
integ_features <- SelectIntegrationFeatures(object.list = seurat_split, nfeatures = 3000)
seurat_split <- PrepSCTIntegration(object.list = seurat_split, anchor.features = integ_features)
```

```{r}
integ_anchors <- FindIntegrationAnchors(object.list = seurat_split, 
                                        normalization.method = "SCT",
                                        anchor.features = integ_features)

seurat_integrated <- IntegrateData(anchorset = integ_anchors,normalization.method = "SCT")
```

