---
title: 'Analysis of Single-Cell RNA-seq - Solutions'
author: "Emily V Chambers"
date: '`r format(Sys.time(), "Last modified: %d %b %Y")`'
output:
  html_notebook:
    toc: yes
    toc_float: yes
    css: stylesheets/styles.css
  pdf_document:
    toc: yes
---
<img src="images/logo-sm.png" style="position:absolute;top:40px;right:10px;" width="200" />

```{r setup, include=FALSE}
knitr::opts_chunk$set(
  tidy = TRUE,
  tidy.opts = list(width.cutoff = 95),
  message = FALSE,
  warning = FALSE,
  fig.width = 10,
  time_it = TRUE
)

```

# Solutions to exercises

# 1

<div class="exercise">
- Use to sampleSheet to determine the directory containing data from the second healthy individual (`PBMMC_2`).
- Use the `Read10X` and `CreateSeuratObject` functions to import the data from this sample, and save as a new variable (`seurat_obj2` for example)
</div>

```{r}
seurat_data <- Read10X(data.dir = "GSE132509/GSM3872443/")
seurat_obj2 <- CreateSeuratObject(counts = seurat_data, 
                                 min.features = 100, 
                                 project = "PBMMC_1")
seurat_obj2

```

# 2

<div class="exercise">
**Exercise:** Now that we have categorized our cells by their cell cycle phase. We can visualise how the cell cycle drives variation in our data by looking at the PCA.

1. Have a go at visualizing the PCA data using a dimensionality reduction plot and colour the cells by their cell cycle phase

**hint** we have used the command for this earlier...
</div>

```{r}
DimPlot(seurat, reduction = "pca", group.by = "Phase")

```
# 3

<div class="exercise">
**Exercise:** 

Lets look at are new UMAP and t-SNE reduction plots. We can use the dimensionality reduction plot `DimPlot` to visualise these. The `reduction=` parameter will need to be changed....


1. Make plots for the UMAP and t-SNE reduced data.
2. See if you can colour the cells by cell phase and sample group (`group.by=`)


</div>

```{r}
DimPlot(seurat_integrated, reduction = "tsne", group.by = "Phase")
DimPlot(seurat_integrated, reduction = "tsne", group.by = "SampleName")

DimPlot(seurat_integrated, reduction = "umap", group.by = "Phase")
DimPlot(seurat_integrated, reduction = "umap", group.by = "SampleName")
```
# 4

<div class="exercise">

**Exercise**

1. Use `FindMarkers` to find differentially expressed marker genes between 2 other clusters of interest
2. Pick a gene at the top of this list and use the following visualisation tools to look at its expression pattern
 * `VlnPlot`
 * `RidgePlot`
 * `FeaturePlot`
 
</div>

```{r, message=FALSE}

mymarkers <- FindMarkers(seurat, ident.1 = "B Cells", ident.2 = "CD20+ B cells", verbose = F)
mymarkers

VlnPlot(seurat, features = "TFEC")
RidgePlot(seurat, features = "TFEC")
FeaturePlot(seurat, features = "TFEC")
```

# 5

<div class="exercise">
**Final Exercise!**

There are a number of different gene set libraries available in `enrichr` https://maayanlab.cloud/Enrichr/#libraries.

1. Have a look for over representation of functional terms in a different cluster of interest (for example Erythrocytes) and check against a different database

</div>
```{r, message=FALSE}

results <- DEenrichRPlot(seurat, ident.1 = "B Cells", enrich.database = "GO_Cellular_Component_2023", max.genes = 1000, num.pathway = 10, verbose=F)
results

```

